File: src/flowmaputility/export/grid_exporter.py
import csv
from pathlib import Path
from typing import TYPE_CHECKING, Iterator

import numpy as np

if TYPE_CHECKING:
    from flowmaputility.grid.info import GridInfo


class GridExporter:
    """
    Экспортирует рассчитанную сетку (скорости газа/жидкости и коды режимов)
    в CSV или XLSX.
    :param grid: Информация о сетке (объект GridInfo)
    :param codes: Матрица кодов режимов
    """

    HEADERS = ("Скорость газа, м/с", "Код режима", "Скорость жидкости, м/с")

    def __init__(self, grid: "GridInfo", codes: np.ndarray) -> None:
        self.grid = grid
        self.codes = codes

    def export(self, path: str) -> None:
        """
        Экспортирует сетку в файл. Формат определяется по расширению пути.
        :param path: Путь до файла (.csv или .xlsx)
        """
        suffix = Path(path).suffix.lower()
        if suffix == ".csv":
            self._to_csv(path)
        elif suffix in (".xlsx", ".xls"):
            self._to_excel(path)
        else:
            raise ValueError(f"Неподдерживаемое расширение файла: {suffix!r}")

    def _rows(self) -> Iterator[tuple[float, int, float]]:
        """
        Возвращает построчно (скорость газа, код режима, скорость жидкости),
        разворачивая 2D-сетку в длинный формат.
        """
        vsg_flat = self.grid.vsg_2d.ravel()
        codes_flat = self.codes.ravel()
        vsl_flat = self.grid.vsl_2d.ravel()
        for vsg, code, vsl in zip(vsg_flat, codes_flat, vsl_flat):
            yield float(vsg), int(code), float(vsl)

    def _to_csv(self, path: str) -> None:
        """
        Сохраняет сетку в CSV (без внешних зависимостей).
        :param path: Путь до .csv файла
        """
        with open(path, "w", newline="", encoding="utf-8-sig") as f:
            writer = csv.writer(f)
            writer.writerow(self.HEADERS)
            writer.writerows(self._rows())

    def _to_excel(self, path: str) -> None:
        """
        Сохраняет сетку в XLSX через openpyxl (опциональная зависимость).
        :param path: Путь до .xlsx файла
        """
        try:
            from openpyxl import Workbook
        except ImportError as e:
            e.add_note(
                "Для экспорта в xlsx нужен пакет openpyxl: "
                "pip install openpyxl (или flowmaputility[excel])"
            )
            raise

        wb = Workbook()
        ws = wb.active
        ws.append(self.HEADERS)  # type: ignore
        for row in self._rows():
            ws.append(row)  # type: ignore
        wb.save(path)

File: app/src/core/gridExporter.js
/**
 * Экспорт рассчитанной сетки (скорости газа/жидкости и коды режимов) в CSV/XLSX.
 * Порт src/flowmaputility/export/grid_exporter.py: там писали файл на диск,
 * здесь - собираем Blob и отдаём браузеру на скачивание.
 *
 * Экспорт в XLSX требует библиотеку SheetJS (глобальный объект `XLSX`),
 * подключаемую в index.html через CDN - см. README.
 */

const HEADERS = ["Скорость газа, м/с", "Код режима", "Скорость жидкости, м/с"];

/**
 * Экранирует поле CSV аналогично Python csv.writer (quoting=QUOTE_MINIMAL по
 * умолчанию): значение оборачивается в кавычки, если содержит запятую,
 * кавычку или перевод строки; внутренние кавычки удваиваются.
 * @param {string | number} value
 * @returns {string}
 */
function csvField(value) {
  const str = String(value);
  if (/[",\r\n]/.test(str)) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * @param {import("./grid.js").GridInfo} grid
 * @param {number[][]} codes
 * @returns {Array<[number, number, number]>} построчно (vsg, code, vsl),
 *   разворачивая 2D-сетку в длинный формат (порядок как у numpy .ravel(): по строкам).
 */
function toRows(grid, codes) {
  const rows = [];
  for (let i = 0; i < grid.resolution; i++) {
    for (let j = 0; j < grid.resolution; j++) {
      rows.push([grid.vsg2d[i][j], codes[i][j], grid.vsl2d[i][j]]);
    }
  }
  return rows;
}

/**
 * @param {string} filename
 * @param {Blob} blob
 */
function downloadBlob(filename, blob) {
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
}

/**
 * Строит текст CSV (с BOM, аналог encoding="utf-8-sig" в Python-версии, чтобы
 * Excel корректно определял кодировку и кириллицу в заголовках). Вынесено
 * отдельно от exportToCsv, чтобы быть тестируемой чистой функцией (без DOM).
 * @param {import("./grid.js").GridInfo} grid
 * @param {number[][]} codes
 * @returns {string}
 */
export function buildCsvText(grid, codes) {
  const rows = toRows(grid, codes);
  const lines = [
    HEADERS.map(csvField).join(","),
    ...rows.map((r) => r.map(csvField).join(",")),
  ];
  return "﻿" + lines.join("\r\n") + "\r\n";
}

/**
 * Экспортирует сетку в CSV и запускает скачивание файла.
 * @param {import("./grid.js").GridInfo} grid
 * @param {number[][]} codes
 * @param {string} [filename]
 */
export function exportToCsv(grid, codes, filename = "flow_map.csv") {
  const csvText = buildCsvText(grid, codes);
  downloadBlob(filename, new Blob([csvText], { type: "text/csv;charset=utf-8" }));
}

/**
 * Экспортирует сетку в XLSX (через глобальный SheetJS `XLSX`) и запускает скачивание.
 * @param {import("./grid.js").GridInfo} grid
 * @param {number[][]} codes
 * @param {string} [filename]
 */
export function exportToXlsx(grid, codes, filename = "flow_map.xlsx") {
  if (typeof XLSX === "undefined") {
    throw new Error(
      "Для экспорта в XLSX нужна библиотека SheetJS - подключите её в index.html"
    );
  }
  const rows = toRows(grid, codes);
  const sheetData = [HEADERS, ...rows];
  const worksheet = XLSX.utils.aoa_to_sheet(sheetData);
  const workbook = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(workbook, worksheet, "Sheet1");
  XLSX.writeFile(workbook, filename);
}

/**
 * Экспортирует сетку в файл, формат определяется по расширению.
 * @param {import("./grid.js").GridInfo} grid
 * @param {number[][]} codes
 * @param {string} filename путь/имя файла (.csv или .xlsx)
 */
export function exportGrid(grid, codes, filename) {
  const suffix = filename.slice(filename.lastIndexOf(".")).toLowerCase();
  if (suffix === ".csv") {
    exportToCsv(grid, codes, filename);
  } else if (suffix === ".xlsx" || suffix === ".xls") {
    exportToXlsx(grid, codes, filename);
  } else {
    throw new Error(`Неподдерживаемое расширение файла: ${suffix}`);
  }
}
